src/azul/service/responseobjects/hca_response_v5.py

```diff
@@ -222,43 +226,135 @@ def _push_content_single_part(self) -> str:
 
         :return: S3 object key
         """
-        parameters = dict(object_key=f'manifests/{uuid4()}.tsv',
-                          data=self._construct_tsv_content().encode(),
-                          content_type='text/tab-separated-values')
+        if self.format == 'tsv':
+            data = self._construct_tsv_content().encode()
+            content_type = 'text/tab-separated-values'
+            object_key = f'manifests/{uuid4()}.tsv'
+        elif self.format == 'bdbag':
+            data = self._construct_bdbag()
+            content_type = 'application/zip'
+            object_key = f'manifests/{uuid4()}.zip'
+        else:
+            assert False
+        parameters = dict(object_key=object_key,
+                          data=data,
+                          content_type=content_type)
         return self.storage_service.put(**parameters)
 
     def _construct_tsv_content(self):
         es_search = self.es_search
 
         output = StringIO()
         writer = csv.writer(output, dialect='excel-tab')
-
         writer.writerow(list(self.manifest_entries['bundles'].keys()) +
                         list(self.manifest_entries['contents.files'].keys()))
         for hit in es_search.scan():
-            self._iterate_hit(hit, writer)
+            self._iterate_hit_tsv(hit, writer)
 
         return output.getvalue()
 
-    def _iterate_hit(self, es_search_hit, writer):
+    def _construct_bdbag(self) -> dict:
+        """Create participant.tsv and sample.tsv and return as zipped BDBag."""
+        es_search = self.es_search
+        sample_tsv = StringIO()
+        sample_writer = csv.writer(sample_tsv, dialect='excel-tab')
+        sample_writer.writerow(
+            list(self.manifest_entries['contents.specimens'].keys()) +
+            list(
+                self.manifest_entries['contents.cell_suspensions'].keys()) +
+            list(self.manifest_entries['bundles'].keys()) +
+            list(self.manifest_entries['contents.files'].keys()) +
+            ['file_url'])
+
+        participant_tsv = StringIO()
+        participant_writer = csv.writer(participant_tsv, dialect='excel-tab')
+        participant_writer.writerow(['entity:participant_id'])
+
+        for hit in es_search.scan():
+            self._iterate_hit_bdbag(hit, sample_writer, participant_writer)
+
+        return self._create_zipped_bdbag({'participant.tsv': participant_tsv, 'sample.tsv': sample_tsv})
+
+    @staticmethod
+    def _create_zipped_bdbag(data: dict):
```

Comment:
If you annotate the argument type, you should also annotate the return value. Instead of dict, use `typing.Mapping` and parameterize by key and value type.

Reply:
I'm still learning about the generic types of the typing module. I tried very `typing.Mapping`, but for now got stuck on how to best handle the iteration in over that structure in `_create_zipped_bag`. I found using `Mapping.Dict` to be the best compromise as (a) it provides finer type granularity compared to `dict`, and (b) retains the ease of iterating over a dictionary. Let me know what you think?

Reply:
> Still puzzled as to how to specify the output type of a `def` if it writes a file disk in the return statement.

This function returns the return value of `archive_bag` which is a path to the file as a `str`, AFAICT. This seem straight-forward, am I missing something?

> I found using `Mapping.Dict` to be the best compromise as (a) it provides finer type granularity compared to `dict`, and (b) retains the ease of iterating over a dictionary. Let me know what you think?

`typing.Mapping` lets you iterate over items. If you need to modify the dict, you should use `typing.MutableMapping`. `typing.Dict` is rarely used. Please show me exactly what prevents you from using `typing.Mapping`. A PyCharm warning? If so which one? Screen shot or Slack call would be appreciated. Replying to comments takes too much of my time.

Reply:
If you've got some time let's discuss the Dict/Mapping (and potentially others). I left it as a Dict for now.